huffman/src/detail/flattened_symbol_bitsize_view.hpp

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include "huffman/src/detail/iterator_interface.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+#include <ranges>
+
+namespace starflate::huffman::detail {
+
+template <std::ranges::view V>
```

Comment:
Sure let's look into that. I'll give it some thought tomorrow or this weekend.

At a high level, we get a range of ranges:
```
[
  (a, b, c), // subrange 1
  (d, e),    // subrange 2
  (f, g, h), // subrange 3
  (i),       // subrange 4
]
```

and we want to flatten to:
```
[a, b, c, d, e, f, g, h, i]
```

This view provides a lazy flatten so we don't need to actually construct the flattened range of all the subranges.

Note that the subranges (`symbol_span`s) don't actually store all the elements but are instead are generators.

We're essentially doing this: https://docs.python.org/3/library/itertools.html#itertools.chain
for a collection of `symbol_span`s.

Reply:
I had some thoughts about this and I'll try to write it out tomorrow.

Reply:
Currently, `table` constructors take a range and potentially a tag, and pass that range to the table_storage's constructor. The table_storage's constructor is responsible for taking that range and filling up the internal container with "valid enough" codes and symbols. Once the table_storage's constructor is complete, no elements can be added or removed, and members such as resize and emplace_back are not exposed upwards. After table_storage's constructor is complete, we get to the function body of table's constructor which is responsible for building the Huffman tree, setting skip fields, and then canonicalizing the codes.

I was thinking about passing the symbol-span+count range down to the table_storage constructor and doing nested for loops, but want to avoid using [raw](https://www.youtube.com/watch?v=W2tWOdzgXHA) [loops](http://www.randomprogramming.com/2014/03/no-raw-loops-1/). And if we're able to reuse existing constructors, then we can focus on reviewing the correctness of the range adapter code, which admittedly requires understanding the range abstraction. But I think this will be pretty useful given that this library is focusing on handling streams of data and we'll likely run into other cases where a range abstraction is useful, even after switching to execution on the GPU.

I've added more documentation to the `flattened_symbol_bitsize_view` class template. I've also removed the `to_code_symbol` function and instead do the equivalent in `flattened_symbol_bitsize_view::iterator::operator*()`. Hopefully things are a bit clearer but let me know if there's anything I can explain or we can also try another approach.

Reply:
The thing that threw me off was the use of concepts to figure out whether to use this flattened_symbol_bitsize_view, and I was suggesting using a constructor tag intead. I was not objecting to the existence of this class.
But I guess from a caller's point of view this is nicer, and you've already figured it out so I guess we can keep it.

Reply:
Sorry I misunderstood. But yeah, I'll generally try to keep the interface simpler for the caller at the expense of more complexity in the implementation.

Reply:
How is your understanding of the view class? I don't mind going over it if there are things you don't understand.